src/stream/src/executor/backfill.rs

```diff
@@ -0,0 +1,380 @@
+                                    break 'backfill_loop;
+                                }
+                                Some(chunk) => {
+                                    // Raise the current position.
+                                    // As snapshot read streams are ordered by pk, so we can
+                                    // just use the last row to update `current_pos`.
+                                    self.current_pos = Some(
+                                        chunk
+                                            .rows()
+                                            .last()
+                                            .unwrap()
+                                            .1
+                                            .row_by_indices(&table_pk_indices),
+                                    );
+
+                                    yield Message::Chunk(Self::mapping_chunk(
+                                        chunk,
+                                        &self.upstream_indices,
+                                    ));
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            let mut finish_on_barrier = |msg: &Message| {
+                if let Some(barrier) = msg.as_barrier() {
+                    self.progress.finish(barrier.epoch.curr);
+                }
+            };
+
+            tracing::debug!(
+                actor = self.actor_id,
+                "Backfill has already finished and forward messages directly to the downstream"
+            );
+
+            // Backfill has already finished.
+            // Forward messages directly to the downstream.
+            let upstream = upstream
+                .map(move |result| result.map(|msg| Self::mapping_message(msg, &upstream_indices)));
+            #[for_await]
+            for msg in upstream {
+                let msg: Message = msg?;
+                finish_on_barrier(&msg);
+                yield msg;
+            }
+        } else {
+            // Forward messages directly to the downstream.
+            let upstream = upstream
+                .map(move |result| result.map(|msg| Self::mapping_message(msg, &upstream_indices)));
+            #[for_await]
+            for message in upstream {
+                yield message?;
+            }
+        }
+    }
+
+    #[expect(clippy::needless_lifetimes, reason = "code generated by try_stream")]
+    #[try_stream(ok = Option<StreamChunk>, error = StreamExecutorError)]
+    async fn snapshot_read(table: &StorageTable<S>, epoch: u64, current_pos: Option<Row>) {
+        // `current_pos` is None means it needs to scan from the beginning, so we use Unbounded to
+        // scan. Otherwise, use Excluded.
+        let range_bounds = if current_pos.is_none() {
+            (Bound::Unbounded, Bound::Unbounded)
+        } else {
+            (Bound::Excluded(current_pos.unwrap()), Bound::Unbounded)
```

Comment:
If the `current_pos` is exactly `\xff\xff..`, we'll get the wrong result here due to the implementation. I think we should fix that first.

Reply:
Actually, I think there is no way for backfill to construct a `current_pos` which finally could be serialized as `\xff\xff`. After investigating our serialize datum method, I found that we always serialize a `u8` for null which is not equal to `\xff`.

Reply:
Good catch! How about adding an assertion for this?

Reply:
Done